src/object/cli_shard.c

```diff
@@ -640,7 +640,7 @@ dc_rw_cb(tse_task_t *task, void *arg)
 	orw = crt_req_get(rw_args->rpc);
 	orwo = crt_reply_get(rw_args->rpc);
 	D_ASSERT(orw != NULL && orwo != NULL);
-	if (ret != 0) {
+	if (ret != 0 && ret != -DER_TX_RESTART) {
```

Comment:
In what case(s) can ret be -DER_TX_RESTART? The comment in this if statement seems to confirm that ret is from crt (when a reply is not available).

Reply:
According to some CI test logs, the client got "-DER_TX_RESART" when handled obj_shard_query_key_cb(). I did not find out who returned such error, but I am sure that just because we missed the dc_tx_op_end() under such case, as to the subsequent daos_tx_restart() hit failure (because invalid tx states). That is why I made the patch. Similar logic as key query case, the same logic is also added here.

Reply:
I'm afraid if we don't figure out how the -DER_TX_RESTARTs ended up in ret, we could miss or hide some bug or change. It would be nicer if we don't have to make the change below.

Reply:
I also cannot find out who set the -DER_TX_RESTART in the task->dt_result. I created another RP with assertion() in related logic to check whether we can reproduce the issue.

https://github.com/daos-stack/daos/pull/4452